README.md
# MGP

## Overview

MGP (<b>M</b>ET <b>G</b>eographic <b>P</b>olygon Library) is a Qt/C++-based tool for processing polygons in a meteorological context.
</b> Polygons are assumed to be defined directly in terms of <a href="https://en.wikipedia.org/wiki/Spherical_coordinate_system">spherical coordinates</a>
and handling of <a href="https://en.wikipedia.org/wiki/Map_projection">map projections</a> is outside the scope of the tool. For finding intersections etc.,
polygon edges are in most cases assumed to follow the <a href="https://en.wikipedia.org/wiki/Great_circle">great circle</a>.

The following use cases are currently supported:

<b>USE CASE 1:</b> Apply a sequence of filters to a set of input polygons and generate a set of output polygons.
Examples of filters include
* <i>south of a latitude</i>,
* <i>northeast of a line defined by two geographical locations</i>, and
* <i>inside a polygon defined by a sequence of three or more geographical locations</i>.

<b>Warning:</b> Currently MGP does not generate the expected result in a couple of special cases, but these are not expected to occur in practice.

<b>USE CASE 2:</b> Generate a SIGMET/AIRMET expression from a sequence of filters.

<b>USE CASE 3:</b> Generate a sequence of filters from a SIGMET/AIRMET expression, and indicate matched and incomplete ranges.


## API/library

The MGP C++ library is located under <code>lib/</code>. The public API offers classes, typedefs and global functions for handling the three use cases listed above.


## MGPView tool

The MGPView tool is located under <code>apps/mgpview/</code>. It is a stand-alone GUI application that demonstrates the capabilities of MGP.


## Installation

<i>ROOT</i> = top level directory of MGP, i.e. where <code>mgp.pro</code> is located.

---------------------------------------------
Compile the source code:
<pre>
cd <i>ROOT</i>
qmake
make
</pre>

Intended result: The <code>mgpview</code> application is generated under <code>apps/mgpview/</code> and
the <code>libmgp.a</code> file is generated under <code>lib/</code>.


---------------------------------------------
Create source code documentation:

<pre>
cd <i>ROOT</i>
doxygen Doxyfile
</pre>

Intended result: HTML documention is generated under ROOT/html/.


debian/changelog
metno-mgp (1.1.20-1) xenial; urgency=medium

  * Removed invalid vertices in result polygons 

 -- Helen Korsmo <helenk@pc4996.pc.met.no>  Mon, 23 Apr 2018 08:17:18 +0200

metno-mgp (1.1.19-2) xenial; urgency=medium

  * adjustments to enable edit dialog in Ted 	
  * compile with qt5

 -- Helen Korsmo <helenk@met.no>  Mon, 27 Nov 2017 14:51:37 +0100

metno-mgp (1.1.19-1) precise; urgency=low

  * Added missing header file

 -- Jo Asplin <jo.asplin@met.no>  Tue, 4 Apr 2017 13:16:00 +0200

metno-mgp (1.1.18-1) precise; urgency=low
 
  * Added missing header file

 -- Jo Asplin <jo.asplin@met.no>  Tue, 4 Apr 2017 12:53:00 +0200

metno-mgp (1.1.17-1) precise; urgency=low

  * Added modal dialog for easy editing of points in WI expression

 -- Jo Asplin <jo.asplin@met.no>  Tue, 4 Apr 2017 05:17:00 +0200

metno-mgp (1.1.16-1) precise; urgency=low

  * Prevent abs lat/lon outside [0, 90] and [0, 180] respectively

 -- Jo Asplin <jo.asplin@met.no>  Mon, 9 Feb 2017 07:48:00 +0200

metno-mgp (1.1.15-1) precise; urgency=low

  * Implemented 'WI only' and 'WI keyword implicit' modes

 -- Jo Asplin <jo.asplin@met.no>  Mon, 10 Jan 2017 14:37:00 +0200

metno-mgp (1.1.14-1) precise; urgency=low

  * Handle certain degenerate cases in polygon intersection
  * Allowed specifying a single point in the xmet expression

 -- Jo Asplin <jo.asplin@met.no>  Mon, 10 Oct 2016 08:43:00 +0200

metno-mgp (1.1.13-1) precise; urgency=low

  * Allowed only single spaces in LINE expressions

 -- Jo Asplin <jo.asplin@met.no>  Mon, 12 Sep 2016 09:17:00 +0200

metno-mgp (1.1.12-1) precise; urgency=low

  * Allowed only single spaces in WI-expression

 -- Jo Asplin <jo.asplin@met.no>  Tue, 30 Aug 2016 06:16:00 +0200

metno-mgp (1.1.11-4) precise; urgency=low

  * new debian package to fix nvidia dependency

 -- Helen Korsmo <helen.korsmo@met.no>  Wed, 17 Aug 2016 12:46:49 +0200

metno-mgp (1.1.11-3) precise; urgency=low

  * debian packages for app and lib

 -- Helen Korsmo <helen.korsmo@met.no>  Wed, 13 Jul 2016 15:20:49 +0200

metno-mgp (1.1.11-2) unstable; urgency=low

    * Fixed small typo

 -- Jo Asplin <jo.asplin@met.no>  Wed, 25 May 2016 13:34:00 +0100

metno-mgp (1.1.11-1) unstable; urgency=low

    * Show intersectable polygons without showing intersection

 -- Jo Asplin <jo.asplin@met.no>  Wed, 25 May 2016 12:26:00 +0100

metno-mgp (1.1.10-1) unstable; urgency=low

    * Added Norwegian municipalities as intersectable polygons

 -- Jo Asplin <jo.asplin@met.no>  Wed, 24 May 2016 13:07:00 +0100

metno-mgp (1.1.9-1) unstable; urgency=low

    * Made FIR matching more specific to minimize risk of false matches
    * Provided supported FIR names in API
    * Always represent zero minutes as '00'
    * Made separating dash mandatory in WI and LINE expressions
    * Made 'AND' between adjacent lat/lon expressions mandatory
    * Added 'WI exclusive' mode
    * Set 'WI exclusive' mode via XMETAreaEdit ctor

 -- Jo Asplin <jo.asplin@met.no>  Wed, 18 May 2016 09:27:00 +0100

metno-mgp (1.1.8-3) unstable; urgency=low

    * fixed bug in latitude filter

 -- Jo Asplin <jo.asplin@met.no>  Mon, 19 Apr 2016 15:22:00 +0100

metno-mgp (1.1.8-2) unstable; urgency=low

    * auto-select supported FIRs only

 -- Jo Asplin <jo.asplin@met.no>  Mon, 19 Apr 2016 13:26:00 +0100

metno-mgp (1.1.8-1) unstable; urgency=low

    * handle FIRs directly in the API
    * auto-select FIR from text edit in MGPView

 -- Jo Asplin <jo.asplin@met.no>  Mon, 19 Apr 2016 09:15:00 +0100

metno-mgp (1.1.7-1) unstable; urgency=low

    * restricted WI filter syntax
    * allowed 'AND' between two lat/lon filters

 -- Jo Asplin <jo.asplin@met.no>  Mon, 12 Apr 2016 12:38:00 +0100

metno-mgp (1.1.6-1) unstable; urgency=low

    * fixed bug in latitude intersection
    * improved highlighting

 -- Jo Asplin <jo.asplin@met.no>  Mon, 4 Apr 2016 06:18:00 +0100

metno-mgp (1.1.5-4) unstable; urgency=low

    * fixed bug

 -- Jo Asplin <jo.asplin@met.no>  Tue, 29 Mar 2016 11:30:00 +0100

metno-mgp (1.1.5-3) unstable; urgency=low

    * more consistent color highlighting

 -- Jo Asplin <jo.asplin@met.no>  Mon, 21 Mar 2016 14:25:00 +0100

metno-mgp (1.1.5-2) unstable; urgency=low

    * set better defaults

 -- Jo Asplin <jo.asplin@met.no>  Mon, 21 Mar 2016 09:32:00 +0100

metno-mgp (1.1.4-2) unstable; urgency=low

    * updated version in main window title

 -- Jo Asplin <jo.asplin@met.no>  Thu, 17 Mar 2016 13:20:00 +0100

metno-mgp (1.1.4-1) unstable; urgency=low

    * added ENOB FIR

 -- Jo Asplin <jo.asplin@met.no>  Thu, 17 Mar 2016 11:44:00 +0100

metno-mgp (1.1.3-1) unstable; urgency=low

    * made sphere editing modes mutually exclusive

 -- Jo Asplin <jo.asplin@met.no>  Tue, 15 Mar 2016 08:51:00 +0100

metno-mgp (1.1.2-1) unstable; urgency=low

    * increased icon contrast

 -- Jo Asplin <jo.asplin@met.no>  Fri, 11 Mar 2016 12:52:00 +0100


metno-mgp (1.1.1-1) unstable; urgency=low

    * added desktop file

 -- Jo Asplin <jo.asplin@met.no>  Thu, 10 Mar 2016 13:50:00 +0100

metno-mgp (1.1.0-1) unstable; urgency=low

    * first package

 -- Alexander Bürger <alexander.buerger@met.no>  Thu, 10 Mar 2016 08:00:08 +0100


lib/mgp.h
#ifndef MGP_H
#define MGP_H

#include <QSharedPointer>
#include <QVector>
#include <QPair>
#include <QList>
#include <QString>
#include <QVariant>
#include <math.h>

#define MGP_BEGIN_NAMESPACE namespace mgp {
#define MGP_END_NAMESPACE }

 /*!
 * \addtogroup mgp mgp
 * This module constitutes the main API.
 * @{
 */

/// \cond
MGP_BEGIN_NAMESPACE
/// \endcond

/**
 * The Point type represents a <a href="https://en.wikipedia.org/wiki/Spherical_coordinate_system">spherical coordinate</a> where
 * the following is assumed by default:
 * <ul>
 * <li> the first component is the longitude (azimuth) in the range [-M_PI, M_PI] </li>
 * <li> the second component is the latitude (inclination) in the range [-M_PI_2, M_PI_2] </li>
 * <li> the radius is 1 (i.e. the point is defined on the unit sphere) </li>
 * </ul>
 * (M_PI and M_PI_2 are defined in /usr/include/math.h)
 *
 * <pre>
 * Examples:
 *   South pole: qMakePair(anyLon, -M_PI_2) (i.e. qMakePair(anyLon,               (-90 / 90) * M_PI_2))
 *   North pole: qMakePair(anyLon,  M_PI_2) (i.e. qMakePair(anyLon,               ( 90 / 90) * M_PI_2))
 *   Oslo:       qMakePair(0.1876, 1.0463)  (i.e. qMakePair((10.75 / 180) * M_PI, (59.95 / 90) * M_PI_2))
 * </pre>
*/
typedef QPair<double, double> Point;
typedef QSharedPointer<QVector<Point> > Polygon;
typedef QSharedPointer<QVector<Polygon> > Polygons;


#define DEG2RAD(d) ((d) / 180.0) * M_PI
#define RAD2DEG(r) ((r) / M_PI) * 180


// --- BEGIN classes --------------------------------------------------

//! Abstract interface for filters.
class FilterBase
{
public:
    virtual ~FilterBase() {}

    enum Type {
        Unsupported,
        POINT, WI,
        E_OF, W_OF, N_OF, S_OF,
        E_OF_LINE, W_OF_LINE, N_OF_LINE, S_OF_LINE,
        NE_OF_LINE, NW_OF_LINE, SE_OF_LINE, SW_OF_LINE
    };
    virtual Type type() const = 0;

    /** Sets the filter value from a QVariant. */
    virtual void setFromVariant(const QVariant &var) = 0;

    /** Returns the filter value as a QVariant. */
    virtual QVariant toVariant() const = 0;

    /** Returns true iff the filter is considered to be in a valid state. */
    virtual bool isValid() const = 0;

    /** Applies the filter.
     * \param inPoly The input polygon (assumed to contain at least three points).
     * \return Zero or more polygons resulting from applying the filter to \c inPoly.
     */
    virtual Polygons apply(const Polygon &inPoly) const = 0;

    /** Returns all intersection points between the filter and the given polygon. */
    virtual QVector<Point> intersections(const Polygon &) const = 0;

    /** Sets the filter state from a SIGMET/AIRMET expression. (### more documentation here!) */
    virtual bool setFromXmetExpr(const QString &, QPair<int, int> *, QPair<int, int> *, QString *) = 0;

    /** Returns the canonical SIGMET/AIRMET expression corresponding to the filter state. */
    virtual QString xmetExpr() const = 0;

    /**
     * Returns true iff the given point is rejected by the filter. A point that is rejected would not be included in any polygon returned
     * by apply() (regardless of the shape of the input polygon), and vice versa: a point that is not rejected will always be included in one of
     * those polygons.
     */
    virtual bool rejected(const Point &) const = 0;
};

//! This filter represent a single point.
class PointFilter : public FilterBase
{
public:
    PointFilter();
    void setPoint(const Point &);
    Point point() const;
    double lon() const { return point_.first; }
    double lat() const { return point_.second; }
protected:
    PointFilter(const Point &);
    Point point_;
private:
    virtual void setFromVariant(const QVariant &);
    virtual QVariant toVariant() const;
    virtual bool isValid() const;
    virtual Type type() const { return POINT; }
    virtual Polygons apply(const Polygon &) const;
    virtual QVector<Point> intersections(const Polygon &inPoly) const;
    virtual bool rejected(const Point &) const;
    virtual bool setFromXmetExpr(const QString &, QPair<int, int> *, QPair<int, int> *, QString *);
    virtual QString xmetExpr() const;
};

//! This is the base class for filters that represent regions explicitly defined as closed polygons.
class PolygonFilter : public FilterBase
{
public:
    void setPolygon(const Polygon &);
    Polygon polygon() const;
protected:
    PolygonFilter();
    PolygonFilter(const Polygon &);
    Polygon polygon_;
private:
    virtual void setFromVariant(const QVariant &);
    virtual QVariant toVariant() const;

    /** Returns true iff the polygon consists of at least three points. */
    virtual bool isValid() const;
};

//! This filter clips away regions outside a specific closed polygon.
class WithinFilter : public PolygonFilter
{
public:
    /** Constructs an object with an empty polygon. */
    WithinFilter(bool = false);

    /** Constructs an object with a given polygon. */
    WithinFilter(const Polygon &, bool = false);

private:
    virtual Type type() const { return WI; }
    virtual Polygons apply(const Polygon &) const;
    virtual QVector<Point> intersections(const Polygon &inPoly) const;
    virtual bool rejected(const Point &) const;
    virtual bool setFromXmetExpr(const QString &, QPair<int, int> *, QPair<int, int> *, QString *);
    virtual QString xmetExpr() const;
    bool keywordImplicit_;
};

//! This filter is currently not implemented, but eventually it is intended to generate the union of two explicitly defined polygons.
class UnionFilter : public PolygonFilter
{
public:
    UnionFilter();
    UnionFilter(const Polygon &);
private:
    virtual Type type() const { return Unsupported; } // ###
    virtual Polygons apply(const Polygon &) const;
    virtual QVector<Point> intersections(const Polygon &inPoly) const;
    virtual bool rejected(const Point &) const;
    virtual bool setFromXmetExpr(const QString &, QPair<int, int> *, QPair<int, int> *, QString *);
    virtual QString xmetExpr() const;
};

//! This is the base class for filters that represent regions that are implicitly defined as being on one or the other side of a line.
class LineFilter : public FilterBase
{
public:
    QString directionName() const;

protected:
    // Returns true and intersection point iff filter intersects great circle arc between given two points. If the filter intersects
    // the arc twice, the intersection closest to the first endpoint is returned.
    virtual bool intersects(const Point &, const Point &, Point *) const = 0;

private:
    virtual Polygons apply(const Polygon &) const;
    virtual QVector<Point> intersections(const Polygon &inPoly) const;
};

//! This filter clips away regions on one or the other side of a specific longitude or latitude.
class LonOrLatFilter : public LineFilter
{
public:
    void setValue(double);
    double value() const;
    bool isLonFilter() const;
protected:
    LonOrLatFilter(double);
    double value_;
private:
    virtual void setFromVariant(const QVariant &);
    virtual QVariant toVariant() const;
    virtual bool setFromXmetExpr(const QString &, QPair<int, int> *, QPair<int, int> *, QString *);
    virtual QString xmetExpr() const;
};

//! This filter clips away regions on one or the other side of a specific longitude.
class LonFilter : public LonOrLatFilter
{
public:
    LonFilter(double);
private:
    virtual bool intersects(const Point &, const Point &, Point *) const;
};

//! This filter clips away regions that are not east of a specific longitude.
class EOfFilter : public LonFilter
{
public:
    EOfFilter(double = DEG2RAD(-8));
private:
    virtual Type type() const { return E_OF; }
    virtual bool isValid() const;
    virtual bool rejected(const Point &) const;
};

//! This filter clips away regions that are not west of a specific longitude.
class WOfFilter : public LonFilter
{
public:
    WOfFilter(double = DEG2RAD(15));
private:
    virtual Type type() const { return W_OF; }
    virtual bool isValid() const;
    virtual bool rejected(const Point &) const;
};

//! This filter clips away regions on one or the other side of a specific latitude.
class LatFilter : public LonOrLatFilter
{
public:
    LatFilter(double);
private:
    virtual Polygons apply(const Polygon &) const;
    virtual QVector<Point> intersections(const Polygon &inPoly) const;
    virtual bool intersects(const Point &, const Point &, Point *) const;
    bool isNOfFilter() const;
};

//! This filter clips away regions that are not north of a specific latitude.
class NOfFilter : public LatFilter
{
public:
    NOfFilter(double = DEG2RAD(60));
private:
    virtual Type type() const { return N_OF; }
    virtual bool isValid() const;
    virtual bool rejected(const Point &) const;
};

//! This filter clips away regions that are not south of a specific latitude.
class SOfFilter : public LatFilter
{
public:
    SOfFilter(double = DEG2RAD(70));
private:
    virtual Type type() const { return S_OF; }
    virtual bool isValid() const;
    virtual bool rejected(const Point &) const;
};

//! This filter clips away regions on one or the other side of a specific line defined by two (lon, lat) points.
class FreeLineFilter : public LineFilter
{
public:
    void setLine(const Point &, const Point &);
    void setLine(const QPair<Point, Point> &);
    void setPoint1(const Point &);
    void setPoint2(const Point &);
    QPair<Point, Point> line() const { return line_; }
    Point point1() const { return line_.first; }
    Point point2() const { return line_.second; }
    double lon1() const { return line_.first.first; }
    double lat1() const { return line_.first.second; }
    double lon2() const { return line_.second.first; }
    double lat2() const { return line_.second.second; }
protected:
    FreeLineFilter(const QPair<Point, Point> &);
    QPair<Point, Point> line_;
private:
    virtual void setFromVariant(const QVariant &);
    virtual QVariant toVariant() const;
    virtual bool isValid() const;
    virtual bool intersects(const Point &, const Point &, Point *) const;
    virtual bool setFromXmetExpr(const QString &, QPair<int, int> *, QPair<int, int> *, QString *);
    virtual QString xmetExpr() const;
    virtual bool rejected(const Point &) const;
};

//! This filter clips away regions that are not east of a specific line.
class EOfLineFilter : public FreeLineFilter
{
public:
    EOfLineFilter(const QPair<Point, Point> & = qMakePair(qMakePair(DEG2RAD(5), DEG2RAD(55)), qMakePair(DEG2RAD(7), DEG2RAD(70))));
private:
    virtual Type type() const { return E_OF_LINE; }
};

//! This filter clips away regions that are not west of a specific line.
class WOfLineFilter : public FreeLineFilter
{
public:
    WOfLineFilter(const QPair<Point, Point> & = qMakePair(qMakePair(DEG2RAD(15), DEG2RAD(55)), qMakePair(DEG2RAD(17), DEG2RAD(70))));
private:
    virtual Type type() const { return W_OF_LINE; }
};

//! This filter clips away regions that are not north of a specific line.
class NOfLineFilter : public FreeLineFilter
{
public:
    NOfLineFilter(const QPair<Point, Point> & = qMakePair(qMakePair(DEG2RAD(-5), DEG2RAD(62)), qMakePair(DEG2RAD(15), DEG2RAD(64))));
private:
    virtual Type type() const { return N_OF_LINE; }
};

//! This filter clips away regions that are not south of a specific line.
class SOfLineFilter : public FreeLineFilter
{
public:
    SOfLineFilter(const QPair<Point, Point> & = qMakePair(qMakePair(DEG2RAD(-5), DEG2RAD(72)), qMakePair(DEG2RAD(15), DEG2RAD(74))));
private:
    virtual Type type() const { return S_OF_LINE; }
};

//! This filter clips away regions that are not northeast of a specific line.
class NEOfLineFilter : public FreeLineFilter
{
public:
    NEOfLineFilter(const QPair<Point, Point> & = qMakePair(qMakePair(DEG2RAD(3), DEG2RAD(53)), qMakePair(DEG2RAD(8), DEG2RAD(72))));
private:
    virtual Type type() const { return NE_OF_LINE; }
};

//! This filter clips away regions that are not northwest of a specific line.
class NWOfLineFilter : public FreeLineFilter
{
public:
    NWOfLineFilter(const QPair<Point, Point> & = qMakePair(qMakePair(DEG2RAD(13), DEG2RAD(53)), qMakePair(DEG2RAD(18), DEG2RAD(72))));
private:
    virtual Type type() const { return NW_OF_LINE; }
};

//! This filter clips away regions that are not southeast of a specific line.
class SEOfLineFilter : public FreeLineFilter
{
public:
    SEOfLineFilter(const QPair<Point, Point> & = qMakePair(qMakePair(DEG2RAD(-3), DEG2RAD(63)), qMakePair(DEG2RAD(18), DEG2RAD(64))));
private:
    virtual Type type() const { return SE_OF_LINE; }
};

//! This filter clips away regions that are not southwest of a specific line.
class SWOfLineFilter : public FreeLineFilter
{
public:
    SWOfLineFilter(const QPair<Point, Point> & = qMakePair(qMakePair(DEG2RAD(-3), DEG2RAD(73)), qMakePair(DEG2RAD(18), DEG2RAD(74))));
private:
    virtual Type type() const { return SW_OF_LINE; }
};


typedef QSharedPointer<FilterBase> Filter;
typedef QSharedPointer<QList<Filter> > Filters;


class FIR
{
public:
    static FIR &instance();

    /**
     * Supported Flight Information Regions
     */
    enum Code {
        Unsupported,
        ENOR, // NORWAY
        ENOB // BODO OCEANIC
    };

    /**
     * Returns supported FIR names.
     */
    QStringList supportedNames() const;

    /**
     * Converts a FIR code to a polygon.
     * @param[in] fir FIR code.
     * @return a non-empty polygon for a supported FIR code, otherwise an empty polygon.
     */
    Polygon polygon(Code fir) const;

    /**
     * Returns the first supported FIR found in a text.
     * @param[in] text Text.
     * @return The code for the first supported FIR found in the text, otherwise the code for an unsupported FIR.
     */
    Code firFromText(const QString &text);

private:
    FIR();

    struct FIRInfo
    {
        Polygon polygon_;
        QString name_;
        FIRInfo() {}
        FIRInfo(const Polygon &polygon, const QString &name)
            : polygon_(polygon)
            , name_(name)
        {}
    };

    QHash<Code, FIRInfo> fir_;
};

// --- END classes --------------------------------------------------


// --- BEGIN global functions --------------------------------------------------

/**
 * Applies a filter sequence to a set of polygons.
 *
 * \param[in] polygons Set of zero or more polygons.
 * \param[in] filters  Sequence of zero or more filters.
 * \return The list of polygons that results from applying \c filters to \c polygons.
 */
Polygons applyFilters(const Polygons &polygons, const Filters &filters);

/**
 * Applies a filter sequence to a single polygon.
 *
 * \note This is an overloaded function.
 * \param[in] polygon Polygon.
 * \param[in] filters Sequence of zero or more filters.
 * \return The list of polygons that results from applying \c filters to \c polygon.
 */
Polygons applyFilters(const Polygon &polygon, const Filters &filters);

/**
 * Converts a filter sequence to a SIGMET/AIRMET area expression.
 *
 * \param[in] filters Sequence of zero or more filters.
 * \return The canonical area part of a SIGMET/AIRMET expression that corresponds to \c filters.
 */
QString xmetExprFromFilters(const Filters &filters);

/**
 * Converts a SIGMET/AIRMET area expression to a filter sequence.
 *
 * \param[in]  expr              SIGMET/AIRMET area expression.
 * \param[out] matchedRanges     Sequence of string position ranges of each fully matched filter in \c expr.
 * \param[out] incompleteRanges  Sequence of string position ranges of each incomplete (partly matched) filter in \c expr along with the reasons why they didn't match.
 * \param[in]  wiExclusive       If true, the WI filter is not allowed to occur together with other types of filters.
 * \param[in]  wiOnly            If true, the WI filter is the only filter allowed. NOTE: wiExclusive is irrelevant if wiOnly is true.
 * \param[in]  wiKeywordImplicit If true, the 'WI' keyword is considered implicit if wiOnly is true. Any occurrence of the 'WI' keyword is then
 *                               considered a syntax error. NOTE: wiKeywordImplicit is irrelevant if wiOnly is false.
 * \return The filter sequence representing fully matched filters in the order of appearance in \c expr.
 */
Filters filtersFromXmetExpr(
        const QString &expr, QList<QPair<int, int> > *matchedRanges = 0, QList<QPair<QPair<int, int>, QString> > *incompleteRanges = 0,
        bool wiExclusive = true, bool wiOnly = true, bool wiKeywordImplicit = false);

/**
 * Returns the SIGMET/AIRMET degrees form of a longitude value in radians ([-M_PI, M_PI]).
 * Examples:
 *        0 -> E00000)
 *    -M_PI -> W18000
 *   M_PI_2 -> E09000
 */
QString xmetFormatLon(double val);

/**
 * Returns the SIGMET/AIRMET degrees form of a latitude value in radians ([-M_PI_2, M_PI_2]).
 * Examples:
 *         0 -> N0000
 *   -M_PI_2 -> S9000
 *    M_PI_4 -> N4500
 */
QString xmetFormatLat(double val);

/**
 * Sets the list of polygons that will be intersected in intersectedPolygons().
 * @param polygons The list of intersectable polygons indexed from 0 to n-1.
 */
void setIntersectablePolygons(const Polygons &polygons);

/**
 * Intersects the polygons last specified in setIntersectablePolygons().
 * @param intersectors The polygons used for intersecting.
 * @return A list consisting of the following for each intersected polygon P:
 *   1) the (zero-based) index of P in the list passed to setIntersectedPolygons()
 *   2) the intersected subpolygons within P
 */
QList<QPair<int, Polygons> > intersectedPolygons(const Polygons &intersectors);

/**
 * Computes the area of a polygon.
 * \param[in] polygon Polygon.
 * \return The area of the polygon in square kilometers.
 */
double area(const Polygon &polygon);

/**
 * Computes the area of a set of polygons.
 * \param[in] polygons Polygons.
 * \return The total area of the polygons in square kilometers.
 */
double area(const Polygons &polygons);

/**
 * Returns the polygons for the Norwegian municipalities.
 */
Polygons norwegianMunicipalities();

// --- END global functions --------------------------------------------------

MGP_END_NAMESPACE
/*! @} end of Doxygen group mgp */

#endif // MGP_H
